### ISSUE TITLE
Proposal: Flag Pools

### ISSUE THREAD

**Issue by reggaemuffin** [SteemCommunity/steem#11]
Currently flags cost the same voting powre that upvotes use. This discourages flags on the basis of missing out on upvotes. To improve the proof of brain protocol, a moderate amount of flags should not punish users. Therefore we propose a flag pools.

A flag pool works similar to how voting power pools work. But the flag power can only be used for flags. If the flag pool is empty, one can pay from the voting power pool again.

We propose to start with 20% flag power, regenerating at 4% per day. This allows for some flagging, but means they are still a valuable resource. And if users need to flag more, they can use their voting power (in cases like steemcleaners).

There were other solutions proposed that have drawbacks we want to highlight.

One idea was to change the rewards curve. But one benefit of a linear curve is the predictability of ones vote which makes for a good user experience and reduces cognitive load of the average user.

Another idea was to have flags have their own pool, potentially scaled by some factor. This has the problem that flag-only users like steemcleaners loose a lot of their power and it actually takes multiple good actors to combat one bad actor. So a change that we do should not take away the ability to do 10 full flags per day.

**Comment by Raycoms** [comment]
Is there a reason it is 4% regeneration per day?
20% flag power means 2 full flags. At 4% regeneration per day, we have less than half a full flag per day. I think 5% regeneration would be easier to understand for the user. (As it is exactly half a full flag per day)

**Comment by reggaemuffin** [comment]
The idea was to keep the 5 day regeneration timeframe to make an implementation easier. Since each full 100% flag costs 2% voting power, this gives 2 flags per day, compared to 10 upvotes.

**Comment by Raycoms** [comment]
Ah understood, it's 2% per full flag, so 4% per day, okay that makes sense. Somehow got confused.

**Comment by TimCliff** [comment]
I suggest trying to keep the proposal simple. A "mirror" of the upvote mana pool for downvotes, that worked equally to upvotes except in the opposite direction is probably the best place to start.

If you start trying to get into discussions and proposals about upvotes or downvotes having to be more or less "expensive" than others, it will add a whole bunch of complexity to the code, as well as likely generate additional resistance from stakeholders depending on whether they think one should be stronger than the other.

**Comment by Raycoms** [comment]
I think that is his idea. Copying the mana pool from upvotes to downvotes but with only 1/5 of the size and 1/5 of the regeneration and only depletes over flags.

**Comment by jonigirl** [comment]
A mirror vp pool for downvotes, would I think encourage more auditing of poor posts, as was mentioned it won't deplete the upvote pool, also dead easy for punters to understand, ones up ones down.

**Comment by iamsmooth** [comment]
I also don't agree with the whole 20% concept, for a variety of reasons. For example:

> If the flag pool is empty, one can pay from the voting power pool again

This is already demonstrated to not work. People do not want to use their valuable upvote power for downvotes. With the 20% rule it means for each 

> additional resistance from stakeholders depending on whether they think one should be stronger than the other

Exactly. If I had to choose I would suggest that downvotes might be cheaper than upvotes. They are generally less dangerous to the ecosystem because _one can not push rewards, directly or indirectly to oneself (or shills/partners, etc.) with downvotes_, unlike upvotes. 

In fact, this is sort of a fanciful brainstorm and forgive the slight off topic here, but consider a hypothetical system where rewards start out divided equally but _only downvotes_ exist. Garbage/abusive content would be downvoted, and the rest would share more or less equally (only content with _no_ downvotes would get the maximum share). 

Would this work? I think it might. Could it be abused and the intended distribution mechanism undermined by selfishly upvoting one's own content? I think not. There would need to be limits on quantity of posting of course to avoid spamming the pool to get a bigger share, and ultimately it might turn out to be unworkable but the idea is illustrative.

On the other hand, the exact parameters of power and recharge rate probably don't impact any implementation much if at all. They are just constants after all.

**Comment by Raycoms** [comment]
Couldn't the relative size of the flag pool depend on some new witness parameter as well? So the witnesses could also decide with a percentage on the size of the reward pool which can improve the tuning and can be used to easily balance the eco-system?

**Comment by iamsmooth** [comment]
@Raycoms Technically it absolutely could (at some cost of making the implementation more complex) although there is some disagreement among witnesses whether more such parameters is actually a good idea for various reasons.

**Comment by reggaemuffin** [comment]
@iamsmooth I don't think that having as many free flags as you have upvotes is a good idea. So the design was geared to incentivising some amount of flagging that could establish a baseline, but not as many flags as there are upvotes. I am still to be convinced otherwise.

Technically, a separate flag pool would be very simple to employ:
```
if vote.is_downvote?
  cost = 0.02 * vote.power //have it cost 2% for one 100% flag, of 20% pool (simplified)
  if pool > cost
    pool -= cost
    make_vote_free()
```

It is basically a filter before a flag happens that check if the downvote pool has enough power. If yes, take from there. If not, take from the normal pool with normal behavior. We already have the mana_bar implementation so it should be fairly easy to implement.

> This is already demonstrated to not work. People do not want to use their valuable upvote power for downvotes. 

I disagree with this notion. Yes, most people don't want to use their upvotes for it. But in cases like steemcleaners it works. With a flag pool we would make it easy to do *some* flags, then harder to do *many* flags but still possible if one forfeits rewards. This would give the users *choice*, which, in my opinion, is an important factor in proof of brain.

**Comment by TimCliff** [comment]
Part of the reason this idea of separate voting power/mana pools hasn’t made it past the idea stage yet is because it is a difficult issue to get consensus on.

If the intention of this is to be more than just an academic exercise and you actually want to get a change into production, I would suggest taking the views of the stakeholders and top 20 witnesses into account. If there isn’t consensus to vote in favor of a change, no matter how good you code it, it will not make it into production.

Personally I would vote against any proposal that makes downvotes less powerful or useful than they are today. In my view, the main purpose of the change we are currently discussing is to do the opposite - make downvoting _more_ effective / useful.

On an unrelated note, any changes that are made at this point should be taking SMTs into account. This will likely make the coding more complicated.

**Comment by reggaemuffin** [comment]
@TimCliff so far many people have been positive about the change in general. So getting consensus is definitely a possibility.

Regarding the effects. A pool like I proposed would make downvotes more powerful. Which is the goal of it in my opinion. 

Regarding SMTs, the voting pools I suggested would translate into SMTs and could even be a parameter that can be configured.

**Comment by iamsmooth** [comment]
> But in cases like steemcleaners it works

Steemcleaners works mostly due to large centralized (and free) ninja-mine delegation from steemit. Prior to that, it always struggled to get support from individual stakeholders, although to be clear it did always get some.

This in no way demonstrates that the current incentives work for downvoting in the sense of a large decentralized pool with mostly community curation/voting, which is what we are trying to accomplish.

There are many reasons why I believe more limited downvotes (ignoring the costly downvotes which I don't believe work at all) than upvotes are no good, including that they perpetuate the anti-downvote stigma (which I believe is critical to break) and the lower number/power of downvotes makes them more likely to attract retaliation, undermining the entire premise.

**Comment by Raycoms** [comment]
I asked myself if flags should be achievable through both pools. Services
as steemcleaners could develop a system similar to vote selling where
people give them their posting authority to help steemcleaners. While
people probably don't want to delegate to them, I doubt that most people
would have an issue with giving them Auth to flag in behalf of them since
it doesn't cost them voting power.
This would make the implementation way more simple as well.
But, I'm still in favor of making the relative size a witness parameter
since this would be a relative simple way for witnesses to counter spam in
certain phases and flag wars and bullying in others.

**Comment by iamsmooth** [comment]
@Raycoms with a large userbase and many subcommunities there will always be flag wars going on in some places, bullying, reward scamming, and all sort of other things. I'm not saying a witness parameter has NO value but I think it is quite limited, and the motivation for it focuses too much on using a GLOBAL parameter to micromanage for it to make sense on a large platform with, say, millions of users and thousands of sub subcommunities (made up numbers ofc).

@reggaemuffin Consider the case of haijin/rancho. That is someone with a million SP account who has been self/team-voting at max vote power (10/day) with no significant consensus support for many months (possibly over a year, I don't remember when it started). To counter that under your 20% rule would require _five_ million SP stakeholders to all apply their downvote power. I feel that is too much. The way I see it, the downvote pool should work the same as the normal upvote pool where if one million SP stakeholder says "yes" and another says "no", there is no payout, but the one who says "no" (putting the rewards back into the pool for the rest of the community to get them) and gets no reward for doing so should not also be charged for it (as is currently the case, which is why the haijin/relaxo scam has gone on so long, along with countless other content-oblivious reward-extracting scams).

You can argue that rancho's 1M SP account is a special case but I disagree. In every corner of the community or subcommunity there are some larger stakeholders relative to most of the rest in that subcommunity, whether that happens to be 1M SP or 1000 SP. It shouldn't take five similarly-large stakeholders (or perhaps 50 with 1/10 stake each) to counter it. One saying yes and one saying no (with equal SP) is enough to demonstrate lack of consensus, and the rewards should flow elsewhere.

**Comment by Raycoms** [comment]
I think in this case we're always discussing a trade-off.
- Should we allow people to easily counter abuse of high staked users.
- Should we allow high staked users to easily bully lower staked users.

Considering the case of haijin/rancho.
- How difficult should it be to counter it?
- How easy do we want to make it so that they can use their power to suppress a lot of users.

If you have it at 1:1 I think it is easier to escalate between two users.
If one whale cancels all upvotes of haejin/rancho. Both can also cancel all upvotes of this whale.

But, when you have a lower percentage. For example 50% it will be more of a consensus decision since it requires at least two whales which also makes it more difficult to counteract.

(At the moment we have it exactly at 1:1 which results in stronger escalations as well in my opinion)

**Comment by iamsmooth** [comment]
The difference is, bullying is annoying and hostile, but it doesn't undermine the fundamental economics of the platform, as reward-extraction schemes are do and are doing. 

I'm not supportive of bullies by any means and I've probably done more than most to counter bullies on the platform (for example, by using my valuable vote power to altruistically repair reputation with direct no reward for me). But let's be realistic: There will always be bullies. Even at 20% rancho (or pick your favorite psychopath) can bully just about anyone he wants.  And if bullies don't downvote, they will bully other ways, like posting nasty things about people. None of this is a serious threat to the viability of the platform the way reward-harvesting is. People complain about trolls and bullies on the internet, but they are complaining about such things on platforms will millions or even billions of users. Those platforms are a great success in spite of the bullying!

When it comes down to it, upvotes are MORE dangerous than downvotes because upvotes can steal rewards for personal gain, and downvotes never can. The idea that downvotes need special restriction is entirely wrong-headed and backwards.

**Comment by Raycoms** [comment]
I agree, good point. Reward generation is what makes Steem unique compared to most other platforms. Bullying will exist independent of this.

**Comment by bobinson** [comment]
@reggaemuffin - quick question that comes to my mind is, will there be any impact to TPS after this change ? How to test it ? 

As per https://www.blocktivity.info/ EOS and BTS has higher transactions per day now and Komodo seems to be catching up. With Casper and the proposed changes in BTC, one of our biggest advantages is the TPS. With EOS + Everipedia combination also live, I will be curious to know how TPS will be impacted if at all and how it will be measured.

**Comment by develCuy** [comment]
@reggaemuffin this is an interesting proposal that requires proper evaluation. Now that we have mana there is better understanding on the "cost" of operations (vote, comment, transfer, claim, delegate). There is no need to have separate mana pools for each of them, so my counter proposal is to have two separate operations: upvote and downvote, both with different mana cost. I think that this is easier to implement and we still encourage counter abuse.

**Comment by Raycoms** [comment]
I think the idea here is not about the usage of cost on the blockchain but
about the cost related to voting power. So there would have to be an
additional pool for down-votes so it does not cost "precious upvote power".


Am Di., 16. Okt. 2018 um 19:34 Uhr schrieb Fernando Paredes García <
notifications@github.com>:

**Comment by TimCliff** [comment]
@develCuy based on your comment, it sounds like there is some confusion. There are currently two completely separate and unrelated "manas"
1. Resource Credit mana
2. Voting mana

Resource Credit mana is the replacement for the "bandwidth" system, and are used to limit a person's transactions.

Voting mana is the replacement for voting power, and is used to determine the weight of someone's vote.

The discussion about separate mana pools in this issue is about the voting mana pools.

**Comment by develCuy** [comment]
@TimCliff thanks for the clarification! Yet, I stand on my proposal, there should be different mana cost for upvotes and downvotes, so they could be different operations. Actually, we could start by having two constants: UPVOTE_MANA_COST & DOWNVOTE_MANA_COST as a proof of concept.

**Comment by TimCliff** [comment]
Are we talking about resource credit mana costs? AFAIK, those are basically just calculations based on the "costs" to the blockchain (state size, cpu, etc.).

**Comment by reggaemuffin** [comment]
@iamsmooth I see your points with bad actors. Having 1:1 pools is a solution that I also could get behind, it would be a lot more interesting in terms of the impact it has on behavior.

**Comment by reggaemuffin** [comment]
@develCuy I don't think flags will be used widespread if they have any cost in upvote mana with them. Which is why we propose to have a pool of downvotes that don't use up upvotes.

@TimCliff completely unrelated to RC :)

**Comment by the-oasis** [comment]
1:1, or a 100% separate downvote pool is overkill and comes with a high social cost which would likely do more harm than good on the platform. We'd just want to flag as little as possible, enough to change the course of any abusive behavior. So an effective-10% separate downvote pool should be low cost enough for the purpose. Also, assuming we're in the new economic equilibrium after the proposed changes (50% curation rewards, n^1.3 reward curve, 10% separate downvote), actors like haejin etc would more likely start using their SP on curation (or curation services) than farming posts, amongst other desired effects on the network.

**Comment by gadrian78** [comment]
I'm neither for nor against this proposal. I can live with whatever community thinks is best for Steem.

I have two big concerns though.

**1. How will a third voting/resource bar look to a regular user?**

Moreover, we will have two voting pools, one for "up" and the other for "down".

**2. Is it likely the uninformed user will think "up" is for positive (or do good) and "down" is for negative (or do harm)**?

Or even less, they will think it's similar to liking or not liking on Youtube? Will a user will a full flagging pool be incentivized to go on a "do harm" campaign? How would this help the situation? Especially if we eventually want more users on Steem, who more likely than not will lean the balance toward the "regular guy".

**Comment by anthonyadavisii** [comment]
@reggaemuffin it's quite an interesting idea. I feel like I do recall a similar idea, if I am not mistaken, that Transisto was talking about. I believe the context was sometime around the Haejin debacle. I think the idea of a separate voting pool has it's advantages but is not without disadvantages.

I think the fundamental question with this approach is would it result in more legitimate flags representing blockchain content moderation or would it result in more vindictive revenge flagging from accounts that had perpetrated abuse such as the well known Haejin / Rancho /Starjuno group against their dissenters. 

Would it likely result in more flags as folks would not feel like they are wasting their VP which could otherwise reap payouts with author and curation rewards? Possibly. But it's the nature of the flags that would tell us whether it would improve proof of brain and we also need to keep in mind that the abusers are going to also have that 20% to retaliate.

I also consider how something like this would affect current off-chain moderation projects, specifically one that rewards flags with upvotes such as @steemflagrewards.  I'm curious how that would affect our ability to reward good blockchain moderating flags, unless of course we receive delegations to meet the demand with potentially 20% more flags from each of our abuser fighters as we hate not having sufficient SP to reward them all (but we do try).

**Comment by TimCliff** [comment]
I'm not sure if this discussion is still active, but I wanted to catch the thread up with other conversations I have had in other places.

Earlier in the issue, I stated:
```
Personally I would vote against any proposal that makes downvotes less powerful or useful than they are today. In my view, the main purpose of the change we are currently discussing is to do the opposite - make downvoting more effective / useful.
```

I have updated my position. I am no longer 100% against a change in this direction.

If someone developed a change where there were separate upvote and downvote mana pools, and downvoting mana depleted 2x as fast as upvote mana (thereby giving each user 2x as much upvoting power as downvoting power) I could support that change.
